### PR TITLE
#1692 - Ministry overaward part 1

### PR DESCRIPTION
**Pull request by dheepak-aot**
## Ministry overaward part 1

- [x] Create ministry API to get overaward balance
- [x]  Create ministry API to get overawards for a student
- [x] Create ministry API to add manual overaward deduction
- [x] Created test cases for all the APIs created
image

- [x] Migration to modify the values of disbursement overaward origin types to match UI and removed `Pending award cancelled`

### CODE REVIEW

File: sources/packages/backend/apps/api/src/route-controllers/overaward/_tests_/e2e/overaward.aest.controller.getOverawardBalance.e2e-spec.ts [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+    await request(app.getHttpServer())
+      .get(endpoint)
+      .auth(
+        await getAESTToken(AESTGroups.BusinessAdministrators),
+        BEARER_AUTH_TYPE,
+      )
+      .expect(HttpStatus.OK)
+      .then((response) => {
+        expect(response.body.overawardBalanceValues.CSLF).toStrictEqual(
+          (400).toFixed(2),
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/route-controllers/overaward/_tests_/e2e/overaward.aest.controller.getOverawardBalance.e2e-spec.ts:99]: Just as a suggestion, the expected value can be checked as `.toBe("400.00")`, I guess.

File: sources/packages/backend/apps/api/src/route-controllers/overaward/_tests_/e2e/overaward.aest.controller.getOverawardBalance.e2e-spec.ts [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+        );
+      });
+  });
+
+  it("Should return an empty response when invalid student id is provided", async () => {
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/route-controllers/overaward/_tests_/e2e/overaward.aest.controller.getOverawardBalance.e2e-spec.ts:104]: If the student does not exist it should return a not found, once the student exists and there is no data, we will return an empty result. Others API endpoints for AEST are using the below check.
```ts
const student = await this.studentService.getStudentById(studentId);
if (!student) {
  throw new NotFoundException("Student not found.");
}
```

File: sources/packages/backend/apps/api/src/route-controllers/overaward/_tests_/e2e/overaward.aest.controller.getOverawardsByStudent.e2e-spec.ts [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+    application.currentAssessment = studentAssessment;
+    await applicationRepo.save(application);
+    // Create an overaward.
+    const reassessmentOveraward = createFakeDisbursementOveraward({ student });
+    reassessmentOveraward.studentAssessment = studentAssessment;
+    reassessmentOveraward.creator = user;
+    reassessmentOveraward.disbursementValueCode = "CSLF";
+    reassessmentOveraward.overawardValue = 500;
+    reassessmentOveraward.originType =
+      DisbursementOverawardOriginType.LegacyOveraward;
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/route-controllers/overaward/_tests_/e2e/overaward.aest.controller.getOverawardsByStudent.e2e-spec.ts:102]: Can we please change the type to be aligned with the overaward created?

File: sources/packages/backend/apps/api/src/route-controllers/overaward/_tests_/e2e/overaward.aest.controller.getOverawardsByStudent.e2e-spec.ts [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+    await request(app.getHttpServer())
+      .get(endpoint)
+      .auth(
+        await getAESTToken(AESTGroups.BusinessAdministrators),
+        BEARER_AUTH_TYPE,
+      )
+      .expect(HttpStatus.OK)
+      .then((response) => {
+        expect(response.body).toHaveLength(1);
+        expect(response.body[0].dateAdded).toBeDefined();
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/route-controllers/overaward/_tests_/e2e/overaward.aest.controller.getOverawardsByStudent.e2e-spec.ts:116]: Can we please have a variable created for `response.body[0]`.
```ts
const [overaward] = response.body;
```

File: sources/packages/backend/apps/api/src/route-controllers/overaward/_tests_/e2e/overaward.aest.controller.getOverawardsByStudent.e2e-spec.ts [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+        expect(response.body[0].overawardOrigin).toBeDefined();
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/route-controllers/overaward/_tests_/e2e/overaward.aest.controller.getOverawardsByStudent.e2e-spec.ts:117]: Please check the expected values.

File: sources/packages/backend/apps/api/src/route-controllers/overaward/models/overaward.dto.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+export class OverawardBalanceAPIOutDTO {
+  overawardBalanceValues: Record<string, number>;
+}
+
+export class OverawardAPIOutDTO {
+  dateAdded: Date;
+  overawardOrigin: DisbursementOverawardOriginType;
+  awardValueCode: string;
+  overawardValue: number;
+  addedByUserFirstName?: string;
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/route-controllers/overaward/models/overaward.dto.ts:20]: Are not usually returning from the API the user name as formatted to be directly used using the method `getUserFullName`?

File: sources/packages/backend/apps/api/src/route-controllers/overaward/models/overaward.dto.ts [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+  addedByUserLastName?: string;
+  applicationNumber?: string;
+  assessmentTriggerType?: AssessmentTriggerType;
+}
+
+export class OverawardManualRecordAPIInDTO {
+  @Length(AWARD_VALUE_CODE_LENGTH)
+  awardValueCode: string;
+  @Min(1)
+  @Max(MAXIMUM_AWARD_VALUE)
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/route-controllers/overaward/models/overaward.dto.ts:30]: We have this const for money values that we have no business direction `MONEY_VALUE_FOR_UNKNOWN_MAX_VALUE`.

File: sources/packages/backend/apps/api/src/route-controllers/overaward/models/overaward.dto.ts [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+  awardValueCode: string;
+  overawardValue: number;
+  addedByUserFirstName?: string;
+  addedByUserLastName?: string;
+  applicationNumber?: string;
+  assessmentTriggerType?: AssessmentTriggerType;
+}
+
+export class OverawardManualRecordAPIInDTO {
+  @Length(AWARD_VALUE_CODE_LENGTH)
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/route-controllers/overaward/models/overaward.dto.ts:27]: Should we validate min/max here? The awards are supposed to be always 4 characters.

File: sources/packages/backend/apps/api/src/route-controllers/overaward/overaward.aest.controller.ts [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+   * @returns overaward balance for student.
+   */
+  @Get("student/:studentId/balance")
+  async getOverawardBalance(
+    @Param("studentId", ParseIntPipe) studentId: number,
+  ): Promise<OverawardBalanceAPIOutDTO> {
+    const overawardBalance =
+      await this.disbursementOverawardService.getOverawardBalance([studentId]);
+
+    return { overawardBalanceValues: overawardBalance[studentId] };
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/route-controllers/overaward/overaward.aest.controller.ts:55]: We may be crossing a line here where we are allowing the services object to be directly returned. If the service class receives more properties it will be directly exposed to the API. Maybe not be a big issue for the AEST but a possible issue when the students will be also receiving the data.

File: sources/packages/backend/apps/api/src/route-controllers/overaward/overaward.aest.controller.ts [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+  @Roles(Role.StudentAddOverawardManual)
+  @Post("student/:studentId")
+  async addManualOverawardDeduction(
+    @Param("studentId", ParseIntPipe) studentId: number,
+    @Body() payload: OverawardManualRecordAPIInDTO,
+    @UserToken() userToken: IUserToken,
+  ): Promise<PrimaryIdentifierAPIOutDTO> {
+    const studentExist = await this.studentService.studentExists(studentId);
+    if (!studentExist) {
+      throw new NotFoundException("Student not found.");
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/route-controllers/overaward/overaward.aest.controller.ts:97]: Missing API annotations.

File: sources/packages/backend/apps/api/src/services/student/student.service.ts [hunk 11]
```diff
@@ -75,7 +75,7 @@ export class StudentService extends RecordDataModelService<Student> {
    * @returns true if the student was found, otherwise false.
    */
   async studentExists(studentId: number): Promise<boolean> {
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/services/student/student.service.ts:77]: Great catch 😉

File: sources/packages/backend/libs/services/src/disbursement-overaward/disbursement-overaward.service.ts [hunk 12]
```diff
@@ -70,4 +75,58 @@ export class DisbursementOverawardService {
...
+        studentAssessment: {
+          id: true,
+          application: { applicationNumber: true },
+          triggerType: true,
+        },
+      },
+      where: {
+        student: { id: studentId },
+      },
+      relations: {
```
Comment (andrewsignori-aot) [inline sources/packages/backend/libs/services/src/disbursement-overaward/disbursement-overaward.service.ts:103]: I would say that we have been configuring the object query following the order as `select`, `relations`, and `where`.

File: sources/packages/backend/libs/services/src/disbursement-overaward/disbursement-overaward.service.ts [hunk 13]
```diff
@@ -70,4 +75,58 @@ export class DisbursementOverawardService {
...
+  async getOverawardsByStudent(
+    studentId: number,
+  ): Promise<DisbursementOveraward[]> {
+    return this.disbursementOverawardRepo.find({
+      select: {
+        createdAt: true,
+        originType: true,
+        overawardValue: true,
+        disbursementValueCode: true,
+        creator: { firstName: true, lastName: true },
```
Comment (andrewsignori-aot) [inline sources/packages/backend/libs/services/src/disbursement-overaward/disbursement-overaward.service.ts:93]: Not part of the PR but we have been not using `creator` for this type of auditing purposes. I guess that this is a miss in the same way that we missed adding the notes here also. Thoughts?

File: sources/packages/backend/libs/services/src/disbursement-overaward/disbursement-overaward.service.ts [hunk 14]
```diff
@@ -70,4 +75,58 @@ export class DisbursementOverawardService {
...
+   * Add a manual overaward deduction.
+   * @param awardValueCode award value code.
+   * @param overawardValueDeducted overaward deducted value.
+   * @param auditUserId user who added overaward deduction
+   * @returns overaward record created.
+   */
+  async addManualOverawardDeduction(
+    awardValueCode: string,
+    overawardValueDeducted: number,
+    studentId: number,
```
Comment (andrewsignori-aot) [inline sources/packages/backend/libs/services/src/disbursement-overaward/disbursement-overaward.service.ts:120]: `studentId` is missing from the comments.

File: sources/packages/backend/libs/services/src/disbursement-overaward/disbursement-overaward.service.ts [hunk 15]
```diff
@@ -70,4 +75,58 @@ export class DisbursementOverawardService {
...
+    auditUserId: number,
+  ): Promise<DisbursementOveraward> {
+    const overawardManualRecord = new DisbursementOveraward();
+    overawardManualRecord.creator = { id: auditUserId } as User;
+    overawardManualRecord.disbursementValueCode = awardValueCode;
+    overawardManualRecord.overawardValue = -overawardValueDeducted;
```
Comment (andrewsignori-aot) [inline sources/packages/backend/libs/services/src/disbursement-overaward/disbursement-overaward.service.ts:126]: I would expect these values to come from the UI ready. The API should just accept it (this would require a change on the class validator as it is right now also).
Either way, while doing it on the UI I believe that we should get the number and multiply it by `-1` to invert it.

File: sources/packages/web/src/services/OverawardService.ts [hunk 16]
```diff
@@ -0,0 +1,56 @@
+import ApiClient from "@/services/http/ApiClient";
+import {
+  OverawardAPIOutDTO,
+  OverawardBalanceAPIOutDTO,
+  OverawardManualRecordAPIInDTO,
+} from "@/services/http/dto";
+
+/**
+ * Client service layer for Student Assessments.
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/services/OverawardService.ts:9]: Please update the comment.

**Review by andrewsignori-aot — COMMENTED** [review]
Great work and pretty cool to see the E2E tests growing. Please take a look at the comments.

File: sources/packages/backend/apps/api/src/route-controllers/overaward/models/overaward.dto.ts (hunk 8 above)
Comment (dheepak-aot) [reply sources/packages/backend/apps/api/src/route-controllers/overaward/models/overaward.dto.ts:27]: Yes, with that assumption only I just used `@length()` alone.

File: sources/packages/backend/apps/api/src/services/student/student.service.ts (hunk 11 above)
Comment (andrepestana-aot) [reply sources/packages/backend/apps/api/src/services/student/student.service.ts:77]: If the `await` was added we are not returning a Promise anymore, right?

File: sources/packages/backend/apps/api/src/route-controllers/overaward/overaward.aest.controller.ts (hunk 9 above)
Comment (dheepak-aot) [reply sources/packages/backend/apps/api/src/route-controllers/overaward/overaward.aest.controller.ts:55]: I get your point here but because of overaward balance being dynamic `Record<string,number>` , how could we control retaining the dynamic behaviour in place? may be for students we need to used typed award codes in DTO?

**Review by andrepestana-aot — COMMENTED** [review]
Great job. Left a comment.

File: sources/packages/backend/libs/services/src/disbursement-overaward/disbursement-overaward.service.ts (hunk 13 above)
Comment (dheepak-aot) [reply sources/packages/backend/libs/services/src/disbursement-overaward/disbursement-overaward.service.ts:93]: I assume you are saying that we must have an exclusive column to identify the user who added the overaward/note right?

**Comment by dheepak-aot** [comment]
> Great job. Left a comment.

On the comment, it still must return promise because it is an async method.

image

File: sources/packages/backend/apps/api/src/route-controllers/overaward/_tests_/e2e/overaward.aest.controller.getOverawardBalance.e2e-spec.ts (hunk 2 above)
Comment (dheepak-aot) [reply sources/packages/backend/apps/api/src/route-controllers/overaward/_tests_/e2e/overaward.aest.controller.getOverawardBalance.e2e-spec.ts:102]: I was following the pattern of API returning a collection of data where we are not used to validating the path param values against the database. 

For e.g. something like this. 
image

Comment (andrewsignori-aot) [reply sources/packages/backend/apps/api/src/route-controllers/overaward/_tests_/e2e/overaward.aest.controller.getOverawardBalance.e2e-spec.ts]: I believe that we have a mixed pattern then, as shown below we have other endpoints doing the check where some of them are also returning a collection. As we talked it seems more accurate to return the `NotFoundException` when the student was not found and the empty collection when the student was found but there is no data.

@Get("student/:studentId")
@Post("student/:studentId")
@Patch(":studentId/save-uploaded-files")
@Get(":studentId/sin-validations")
@Post(":studentId/sin-validations")
@Patch(":studentId/sin-validations/:sinValidationId")
@Patch(":studentId/sin-validations/:sinValidationId")
@Post(":studentId/files")
@Get(":studentId/application-summary")

File: sources/packages/backend/apps/api/src/route-controllers/overaward/models/overaward.dto.ts (hunk 8 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/api/src/route-controllers/overaward/models/overaward.dto.ts]: I saw the changed code 😉

File: sources/packages/backend/apps/api/src/route-controllers/overaward/overaward.aest.controller.ts (hunk 9 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/api/src/route-controllers/overaward/overaward.aest.controller.ts:63]: As we talked this is a key/value pair that does not seem likely to be expanded. I would not have a major concern to keep it as is for AEST and Ministry.

File: sources/packages/backend/apps/api/src/route-controllers/overaward/_tests_/e2e/overaward.aest.controller.getOverawardBalance.e2e-spec.ts [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+      .then((response) => {
+        expect(response.body.overawardBalanceValues.CSLF).toBe("400.00");
+      });
+  });
+
+  it("Should throw not found error when invalid student id is provided", async () => {
+    // Arrange
+    const student = await studentRepo.save(createFakeStudent());
+    // Create an overaward.
+    const legacyOveraward = createFakeDisbursementOveraward({ student });
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/route-controllers/overaward/_tests_/e2e/overaward.aest.controller.getOverawardBalance.e2e-spec.ts:106]: I believe that we can skip the overaward creation here since the method will stop at the student not found exception. Does it make sense?

**Review by andrewsignori-aot — APPROVED** [review]
Thanks for doing the changes. I left a comment on one E2E test but the PR looks good 👍

File: sources/packages/backend/libs/sims-db/src/entities/disbursement-overaward-origin.type.ts [hunk 18]
```diff
@@ -10,20 +10,14 @@ export enum DisbursementOverawardOriginType {
...
    * Part of an overaward balance was used to change an award value.
    */
-  AwardValueAdjusted = "Award value adjusted",
-  /**
-   * An pending award that included an overaward was cancelled and
-   * the overaward value must to be added back to the overaward balance.
-   * @deprecated to be removed.
-   */
-  PendingAwardCancelled = "Pending award cancelled",
+  AwardDeducted = "Award deducted",
```
Comment (andrepestana-aot) [inline sources/packages/backend/libs/sims-db/src/entities/disbursement-overaward-origin.type.ts:13]: I think we need to update the comment as we are more specific here to 'deducted' now. Also, are we going to to have an "AwardAdded"?

**Review by andrepestana-aot — APPROVED** [review]
Great job. I just left another question.

File: sources/packages/backend/libs/sims-db/src/entities/disbursement-overaward-origin.type.ts (hunk 18 above)
Comment (dheepak-aot) [reply sources/packages/backend/libs/sims-db/src/entities/disbursement-overaward-origin.type.ts:13]: IMO the existing comment on AwardDeducted(this is award deducted to pay the overaward balance) seems to be fine. We can adjust if required in upcoming PR.
